src/finch/tensor.py

```diff
@@ -587,25 +609,37 @@ def asarray(obj, /, *, dtype=None, format=None):
         return tensor
 
 
+def reshape(x: Tensor, /, shape: tuple[int, ...], *, copy: bool | None = None) -> Tensor:
+    # Only to run array-api-tests that require it for multiple tests. Maybe raise a warning
+    # that it densifies? Or enable when some env var is present?
+    arr = x.todense()
+    arr = arr.reshape(shape)
+    return Tensor(arr)
+
+
 def full(
     shape: int | tuple[int, ...],
     fill_value: jl_dtypes.number,
     *,
     dtype: DType | None = None,
     format: str = "coo",
 ) -> Tensor:
+    if not np.isscalar(fill_value):
+        raise ValueError("`fill_value` must be a scalar")
+    if format not in ("coo", "dense"):
+        raise ValueError(f"{format} format not supported.")
     if isinstance(shape, int):
         shape = (shape,)
-    if dtype is not None:
-        fill_value = dtype(fill_value)
+    dtype = np.asarray(fill_value).dtype.type if dtype is None else jl_dtypes.jl_to_np_dtype[dtype]
+    if dtype == np.bool_:  # Fails with: Finch currently only supports isbits defaults
+        dtype = bool
 
-    if format == "coo":
+    if format == "coo" and shape != ():
         return Tensor(
-            jl.Tensor(jl.SparseCOO[len(shape)](jl.Element(fill_value)), *shape)
+            jl.Tensor(jl.SparseCOO[len(shape)](jl.Element(dtype(fill_value))), *shape)
         )
-    if format == "dense":
-        return Tensor(np.full(shape, fill_value, dtype=jl_dtypes.jl_to_np_dtype[dtype]))
-    raise ValueError(f"{format} format not supported.")
+    else:
+        return Tensor(np.full(shape, fill_value, dtype=dtype))
```

Comment:
This returns a fully dense tensor. Can we return an empty tensor with the right fill-value?

Reply:
only if the fill value is "special" right? we wouldn't want to compile a ton of kernels on accident.

Reply:
There are two formats supported by `full` here: `coo` and `dense`. The `else` branch is for `dense` case so I think we should return an actual dense tensor. Or do we want a COO format?

Reply:
in Julia and Matlab we have `spzeros` to explicitly ask for a sparse array of zeros, as opposed to `zeros` which returns a dense array. I think this would be resolved if we simply require the user to type `dense` or `coo` and the else case is an error. That way, the user can't make a dense tensor on accident.

Reply:
Oh, I see, we check it at the top. In that case, I think this is fine.

Reply:
We could run into a performance cliff if we choose the incorrect default either way. I think the right solution would be the middle solution with a dynamic fill-value, treating it as a runtime constant rather than a compile time constant, except for absorbing or identity elements.

Reply:
Finch doesn't support dynamic sparse fill values yet.

Reply:
I could add it with the ISO level?

Reply:
but it's not a common thing to do

Reply:
Hmm. 🤔 Not an easy decision either way. Let's pick any solution here and create a Todo comment and an issue to fix it up later?

Reply:
I'm okay with the code as written, though I think it could be improved to make it clearer that dense is not the default but an intentional user decision.

Reply:
The issue with that is we're introducing finch-specific extensions to the standard, and they're highly unlikely to get used.

Reply:
which extension here is finch-specific? `full`? or the `format` keyword?

Reply:
I think nothing is finch-specific. `full` follows Array API with and additional keyword only `format` argument that is also present in `full` in Numba backend.

Reply:
If you could change the `else` to an `elif format == dense`, I think I approve this PR.

Reply:
`format` is sparse-specific.

Reply:
I updated `if` statement and kept the `format` keyword for this PR but if you think it should be dropped as it's sparse-specific, let me know - I will keep only COO then. (but Numba backend also has `format` keyword for these functions and accepts for example "gcxs")

Reply:
No, I think adding a format keyword is becoming a good, uniform way to incorporate sparsity into these functions without changing the signatures too much, let's keep doing it!

Reply:
I would add `format` to `compute`, to minimise the amount of additional API added. Uniformity is good, less API to do the same thing is even better.

Reply:
I think it's very reasonable for a user to want an empty COO matrix full of ones.

Reply:
so they would type `full(1, format=coo)`. You're proposing that they need to do `compute(lazy(full(1)), format=dense)`?

Reply:
or would it be `compute(full(lazy(1)), format="coo")`? I'm not quite sure where the lazy would go in this invocation. Another issue is that we may want to specify intermediate formats for performance reasons, so it's not just the final format that matters. Finally, it would be nice to be able to change the format without dropping into the lazy interface.

Reply:
I think it would be close to `full(1).compute(format[, optional_schedule])` in a lazy context (i.e. inside a function with `@compiled`.

`compute` would force materialisation in any context.

I guess a middle way would be to add it to creation functions only, but every function would be way too much API surface.

Reply:
I'm not sure we can support a lazy version of this function. I agree that adding `format` to all creation functions is a good plan. As a future API issue, we should consider whether to add format to eager-only operations.